pkg/flowaggregator/s3uploader/s3uploader.go

```diff
@@ -0,0 +1,638 @@
+			}
+			if err == nil {
+				committedRec += committed
+			}
+			if s3.commitTicker != nil {
+				s3.mutex.Lock()
+				s3.commitTicker.Reset(s3.commitInterval)
+				s3.mutex.Unlock()
+			}
+		case <-logTicker.C:
+			klog.V(4).InfoS("Total number of records committed to S3", "count", committedRec)
+			committedRec = 0
+		}
+	}
+}
+
+// batchCommitAll commits all flow records cached in local deque in one uploadFile
+// call. Returns the number of records successfully committed, and error if
+// encountered. Cached records will be removed only after a successful commit.
+func (s3 *S3ExportProcess) batchCommitAll(ctx context.Context, uploaderAPI S3UploaderAPI) (int, error) {
+	// populate records from deque
+	s3.dequeMutex.Lock()
+	currSize := s3.deque.Len()
+	if currSize == 0 {
+		s3.dequeMutex.Unlock()
+		return 0, nil
+	}
+	// Clear the previous records in expiredRecords slice. Allocated memory is kept.
+	s3.expiredRecords = s3.expiredRecords[:0]
+	for i := 0; i < currSize; i++ {
+		record, ok := s3.deque.PopFront().(*flowrecord.FlowRecord)
+		if !ok {
+			continue
+		}
+		s3.expiredRecords = append(s3.expiredRecords, record)
+	}
+	s3.dequeMutex.Unlock()
+
+	buffer.Reset()
+	err := s3.writeFile(&buffer)
+	if err != nil {
+		return 0, err
+	}
+	err = s3.uploadFile(ctx, uploaderAPI, &buffer)
+	if err != nil {
+		// LeavePartsOnError of s3.Uploader indicates whether to leave
+		// successfully uploaded parts in S3, with false as the default
+		// value.
+		s3.pushRecordsToFrontOfQueue(s3.expiredRecords)
+		return 0, err
+	}
+	return len(s3.expiredRecords), nil
+}
+
+func (s3 *S3ExportProcess) writeFile(writer io.Writer) error {
+	if *s3.compress {
+		gzipWriter := gzip.NewWriter(writer)
+		writer = gzipWriter
+		defer gzipWriter.Close()
+	}
+	for _, record := range s3.expiredRecords {
+		err := asCSV(writer, record)
+		if err != nil {
+			return err
+		}
+		io.WriteString(writer, "\n")
+	}
+	return nil
+}
+
+func (s3 *S3ExportProcess) uploadFile(ctx context.Context, uploaderAPI S3UploaderAPI, reader io.Reader) error {
+	fileName := fmt.Sprintf("records-%s.csv", getToken(12))
+	if *s3.compress {
+		fileName += ".gz"
+	}
+	key := fileName
+	if s3.bucketPrefix != "" {
+		key = fmt.Sprintf("%s/%s", s3.bucketPrefix, fileName)
+	}
+	if _, err := uploaderAPI.Upload(ctx, &s3service.PutObjectInput{
+		Bucket: aws.String(s3.bucketName),
+		Key:    aws.String(key),
+		Body:   reader,
+	}); err != nil {
+		return fmt.Errorf("error when uploading file to S3: %v", err)
+	}
+	return nil
+}
+
+func getToken(length int) string {
+	token := ""
+	codeAlphabet := "abcdefghijklmnopqrstuvwxyz"
+	codeAlphabet += "0123456789"
+
+	for i := 0; i < length; i++ {
+		token += string(codeAlphabet[randSeq(int64(len(codeAlphabet)))])
+	}
+	return token
+}
+
+func randSeq(max int64) int64 {
+	// #nosec G404: random number generator not used for security purposes
+	nBig, _ := rand.Int(rand.Reader, big.NewInt(max))
+	return nBig.Int64()
+}
```

Comment:
I don't understand this function:
- why is it called `randSeq`?
- why is the input parameter an `int64` and not an `int`?
- why is the returned a value an `int64` and not an `int`?
- why not just use `rand.Intn`?